packages/core/logger/src/index.js

```diff
@@ -18,152 +18,169 @@ export const LogLevels = {
-      this._logFunction(...args.filter(arg => arg !== undefined && arg !== null).map(arg => {
+      this._logFunction(...args.filter(arg => arg ?? false).map(arg => {
         if (color) {
           return typeof arg === 'object' ? util.inspect(arg, 2)[color] : arg[color];
         }
         return typeof arg === 'object' ? util.inspect(arg, 2) : arg;
       }));
     };
     this.logFile = options.logFile;
-    this.context = options.context;
+
+    this.writeToFile = async.cargo((tasks, callback) => {
+      if (!this.logFile) {
+        return callback();
+      }
+      let logs = '';
+      let origin = "[" + ((new Error().stack).split("at ")[3]).trim() + "]";
+      tasks.forEach(task => {
+        logs += `[${date.format(new Date(), DATE_FORMAT)}] ${task.prefix} ${task.args}\n`;
+      });
+      this.fs.appendFile(this.logFile, `\n${origin} ${logs}`, err => {
+        if (err) {
+          this.logFunction(`There was an error writing to the log file: ${err}`, 'red');
+          return callback(err);
+        }
+        callback();
+      });
+    });
+  }
+
+  registerAPICall(plugins) {
+    let plugin = plugins.createPlugin('logger', {});
+    plugin.registerAPICall(
+      'ws',
+      '/embark-api/logs',
+      (ws, _req) => {
+        this.events.on("log", (logLevel, logMsg) => {
+          logMsg = escapeHtml(logMsg);
+          ws.send(JSON.stringify({msg: logMsg, msg_clear: logMsg.stripColors, logLevel: logLevel}), () => {});
+        });
+      }
+    );
   }
 
-  /**
-   * Parses the logFile, returning an array of JSON objects containing the
-   * log messages.
-   * @param {Number} limit specifies how many log messages to return from the
-   *  end of the log file
-   * @returns {Array} array containing
-   *  - msg: the log message
-   *  - logLevel: log level (ie 'info', 'debug')
-   *  - name: process name (always "embark")
-   *  - timestamp: timestamp of log message (milliseconds since 1/1/1970)
-   */
-  parseLogFile(limit) {
-    let matches;
-    let logs = [];
-    const logFile = fs.readFileSync(this.logFile, 'utf8');
-    while ((matches = LOG_REGEX.exec(logFile)) !== null) {
-     // This is necessary to avoid infinite loops with zero-width matches
-     if (matches.index === LOG_REGEX.lastIndex) {
-       LOG_REGEX.lastIndex++;
-     }
-
-     if (matches && matches.length) {
-       logs.push({
-         msg: [matches[3]],
-         logLevel: matches[2],
-         name: 'embark',
-         timestamp: date.parse(matches[1], DATE_FORMAT).getTime()
-       });
-     }
+  error(...args) {
+    if (!args.length || !(this.shouldLog('error'))) {
+      return;
     }
 
-    // if 'limit' is specified, get log lines from the end of the log file
-    if(limit && limit > 0 && logs.length > limit){
-      logs.slice(limit * -1);
+    let callback = () => {};
+    if (typeof args[args.length - 1] === 'function') {
+      callback = args[args.length - 1];
+      args.splice(args.length - 1, 1);
```

Comment:
The `callback` needs to be determined before we pass `args` onto the `logFunction`, that's why this needs to be done in every log API (unfortunately)